File: 1001-2000/1488. Avoid Flood in The City.java
// Medium
// TreeSet, Map, Greedy
// Time:O(nlogn),Space:O(n)
// https://leetcode.cn/problems/avoid-flood-in-the-city

import java.util.*;

// We want to find the next rain day for the same lake.

class Solution {
    public int[] avoidFlood(int[] rains) {
        Map<Integer, Integer> map = new HashMap<>(); // full lake -> days
        TreeSet<Integer> dry = new TreeSet<>();
        int[] ans = new int[rains.length];

        for (int i = 0; i < rains.length; i++) {
            if (rains[i] == 0) {
                dry.add(i);
                ans[i] = 1;
            } else {
                int lake = rains[i];
                ans[i] = -1;

                if (map.containsKey(lake)) {
                    Integer sunnyDay = dry.ceiling(map.get(lake));

                    if (sunnyDay == null) {
                        // which means there is no sunny days ->flood
                        return new int[0];
                    }
                    ans[sunnyDay] = lake;
                    dry.remove(sunnyDay);
                }
                map.put(lake, i);
            }
        }
        return ans;
    }
}

File: 1001-2000/1488. Avoid Flood in The City.py
# Time:O(nlogn), Space:O(n)

from typing import List

class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        n = len(rains)
        res = [-1] * n
        full_lakes = {}  # (lake no.:prev raining day)
        dry_days = []

        for i in range(n):
            if rains[i] > 0:
                lake = rains[i]

                if lake in full_lakes:
                    # binary search find first > full_lakes[lake]
                    left, right = 0, len(dry_days)
                    while left < right:
                        mid = (left + right) // 2
                        if dry_days[mid] > full_lakes[lake]:
                            right = mid
                        else:
                            left = mid + 1

                    if left == len(dry_days):
                        return []

                    day = dry_days[left]
                    res[day] = lake
                    dry_days.pop(left)

                full_lakes[lake] = i

            else:
                dry_days.append(i)
                res[i] = 1

        return res
